network/simple-onchain-discovery/src/discovery.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+/// A listener for changes on an incoming discovery source.  The entire set of discovered peers must
+/// be provided in each update.
+#[async_trait]
+pub trait DiscoveryListener: Stream<Item = HashMap<PeerId, Peer>> + Unpin {
+    /// `NetworkContext` for logging
+    fn network_context(&self) -> &NetworkContext;
+
+    /// The `DiscoverySource` for purposes of keeping track of each separately
+    fn discovery_source(&self) -> DiscoverySource;
+
+    fn update_channel(&mut self) -> &mut channel::Sender<ConnectivityRequest>;
+
+    async fn start(mut self: Box<Self>) {
+        info!(
+            NetworkSchema::new(self.network_context()),
+            "{} Starting {} Discovery",
+            self.network_context(),
+            self.discovery_source()
+        );
+
+        while let Some(update) = self.next().await {
+            info!("{} Sending update: {:?}", self.network_context(), update);
+            let request =
+                ConnectivityRequest::UpdateDiscoveredPeers(self.discovery_source(), update);
+            if let Err(error) = self.update_channel().try_send(request) {
+                error!(
+                    "{} Failed to send update {:?}",
+                    self.network_context(),
+                    error
+                );
+            }
+        }
+    }
+}
+
+/// Watcher for a file containing discovered peers at a given interval
+pub struct FileChangeListener {
+    network_context: Arc<NetworkContext>,
+    file_path: PathBuf,
+    update_channel: channel::Sender<ConnectivityRequest>,
+    check_interval: Duration,
+    time_service: TimeService,
+}
+
+impl FileChangeListener {
+    pub fn new(
+        network_context: Arc<NetworkContext>,
+        file_path: &Path,
+        update_channel: channel::Sender<ConnectivityRequest>,
+        check_interval: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileChangeListener {
+            network_context,
+            file_path: file_path.to_path_buf(),
+            update_channel,
+            check_interval,
+            time_service,
+        }
+    }
+
+    /// Loads the YAML configuration file
+    fn load_file(&self) -> Result<HashMap<PeerId, Peer>> {
+        let contents = std::fs::read_to_string(&self.file_path).map_err(Error::File)?;
+        serde_yaml::from_str(&contents).map_err(|err| Error::Parsing(err.to_string()))
+    }
+}
+
+#[async_trait]
+impl DiscoveryListener for FileChangeListener {
+    fn network_context(&self) -> &NetworkContext {
+        &self.network_context
+    }
+
+    fn discovery_source(&self) -> DiscoverySource {
+        DiscoverySource::File
+    }
+
+    fn update_channel(&mut self) -> &mut Sender<ConnectivityRequest> {
+        &mut self.update_channel
+    }
+}
+
+impl<'a> Stream for FileChangeListener {
+    type Item = HashMap<AccountAddress, Peer>;
+
+    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.load_file() {
```

Comment:
yeah, probably just store a time_service::Interval in the FileChangeListener and poll the interval here. when it's Poll::Ready, you load the file